### PR TITLE
Fix if expressions in lock expression

### PR DESCRIPTION
**Pull request by crthpl**

### CODE REVIEW

File: vlib/sync/array_rlock_test.v [hunk 1]
```diff
@@ -1,5 +1,5 @@
 fn test_shared_modification() {
-	shared foo := &[2, 0, 5]
+	shared foo := [2, 0, 5]
```
Comment (medvednikov) [inline vlib/sync/array_rlock_test.v:2]: the removals of & seem to be unrelated to the PR?

Comment (crthpl) [reply vlib/sync/array_rlock_test.v:2]: `shared` implies a reference and the underlying type looks like this;
```c
struct __shared__Array_int {
	sync__RwMutex mtx;
	Array_int val;
};
```
and then whenever you creacte a `shared foo`, you create a reference to `__shared__Array_int`, not the `Array_int`:
```c
__shared__Array_int* foo = (__shared__Array_int*)__dup_shared_array(&(__shared__Array_int){.mtx = {0}, .val =
new_array_from_c_array(3, 3, sizeof(int), _MOV((int[3]){2, 0, 5}))}, sizeof(__shared__Array_int));
```
This was generated with current V, and it didn't generate a reference to the array anyways.
This PR makes assignments always use `expr_with_cast`, which doesn't allow casting a reference to shared: https://github.com/vlang/v/blob/cee7856c0f52da2470e4f6c7e53b1f9a0e87d5ee/vlib/v/gen/c/cgen.v#L2212
You need a reference to `__shared__Array_int` because the mutexes all need to be the same one in memory, and having a shared reference is usually (always?) never what you want.

Comment (spytheman) [reply vlib/sync/array_rlock_test.v:2]: This cgen error is not triggered by any of the existing tests, and the tests that are changed, compile fine on both master and the branch of this PR.

Comment (spytheman) [reply vlib/sync/array_rlock_test.v:2]: The file `vlib/v/tests/shared_if_expr_test.v` is the only test file that should stay in the PR imho.

Comment (spytheman) [reply vlib/sync/array_rlock_test.v:2]: For example the change in `vlib/sync/array_rlock_test.v` is not related to the PR:
image

Comment (medvednikov) [reply vlib/sync/array_rlock_test.v:2]: @spytheman he's removing unnecessary `&[]` from the code. It's already handled by V when it sees it's `shared`.

Comment (spytheman) [reply vlib/sync/array_rlock_test.v:2]: I understand that. 
But it is handled by V *now*. 
It is not guaranteed to be handled by V *in the future*. 
By not changing these tests, it would be, because they will break and the CI will tell us when that happens.

Comment (spytheman) [reply vlib/sync/array_rlock_test.v:2]: btw, I have no objections to making it an explicit checker error, and a corresponding test for it.

Comment (medvednikov) [reply vlib/sync/array_rlock_test.v:2]: I see, ok I agree.
